eclipse-theia/theia
# Fix memory leak in plugin tree views

Fixes #10404.

#### What it does
Fixes a bug in `TreeViewExtImpl<T>.getChildren(parentId: string)`, where newly created top-level children were added to the old root node instead of the new one. This caused a memory leak by retaining an implicit reference to the old root node from the new one and failing to properly track and, hence, dispose the top-level children.

The fix ensures that old nodes get disposed and become subject to garbage collection by reusing the single root node instead of creating a new one on full refresh.

See https://github.com/eclipse-theia/theia/issues/10404#issuecomment-1481293316 for more information.

#### How to test
Verify that plugin tree views still work as expected, and there is no memory leak on full refresh now.

For example, open Theia on the [vscode-extension-samples/helloworld-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/helloworld-sample) workspace, and use the `Node Dependencies` view contributed by the [vscode-extension-samples/tree-view-sample](https://github.com/microsoft/vscode-extension-samples/tree/main/tree-view-sample):

* Take a heap snapshot using Chrome DevTools and verify that there is exactly six `Dependency` objects.
* Refresh the view and take another heap snapshot. Verify that there is still exactly six `Dependency` objects.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
